client/cli/src/runner.rs

```diff
@@ -259,24 +264,25 @@ impl<C: SubstrateCli> Runner<C> {
 		}
 	}
 
-	fn run_service_until_exit<T, F>(mut self, service_builder: F) -> Result<()>
-	where
-		F: FnOnce(Configuration) -> std::result::Result<T, sc_service::error::Error>,
-		T: AbstractService + Unpin,
-	{
-		let service = service_builder(self.config)?;
-
-		// we eagerly drop the service so that the internal exit future is fired,
-		// but we need to keep holding a reference to the global telemetry guard
-		// and drop the runtime first.
-		let _telemetry = service.telemetry();
-
-		// we hold a reference to the base path so if the base path is a temporary directory it will
-		// not be deleted before the tokio runtime finish to clean up
-		let _base_path = service.base_path();
+	fn run_service_until_exit(
+		mut self,
+		initialise: impl FnOnce(Configuration) -> sc_service::error::Result<KeepAliveChainComponents>,
+	) -> Result<()> {
+		let KeepAliveChainComponents {
+			task_manager,
+			// we eagerly drop the service so that the internal exit future is fired,
+			// but we need to keep holding a reference to the global telemetry guard
+			// and drop the runtime first.
+			telemetry: _telemetry,
+			// we hold a reference to the base path so if the base path is a temporary directory it will
+			// not be deleted before the tokio runtime finish to clean up
+			base_path: _base_path,
+			rpc: _rpc,
+			..
+		} = initialise(self.config)?;
```

Comment:
I wonder if it would be possible to use `Any` to store the keep-alive-thingies that are not used explicitly. The initialize's return type should be something like `(TaskManager, Box<Any>)` so sc-cli won't need to know what's in there.

Reply:
(It probably don't need to know it's a taskmanager either. It could be just a future)

Reply:
Oh yeah, that's a good idea.